pennylane/utils.py

```diff
@@ -98,5 +109,140 @@ def unflatten(flat, model):
+    """
+    if num_wires == 1:
+        # total number of wires is 1, simply return the matrix
+        return U
+
+    N = num_wires
+    wires = np.asarray(wires)
+
+    if np.any(wires < 0) or np.any(wires >= N) or len(set(wires)) != len(wires):
+        raise ValueError("Invalid target subsystems provided in 'wires' argument.")
+
+    if U.shape != (2 ** len(wires), 2 ** len(wires)):
+        raise ValueError("Matrix parameter must be of size (2**len(wires), 2**len(wires))")
+
+    # generate N qubit basis states via the cartesian product
+    tuples = np.array(list(itertools.product([0, 1], repeat=N)))
+
+    # wires not acted on by the operator
+    inactive_wires = list(set(range(N)) - set(wires))
+
+    # expand U to act on the entire system
+    U = np.kron(U, np.identity(2 ** len(inactive_wires)))
+
+    # move active wires to beginning of the list of wires
+    rearranged_wires = np.array(list(wires) + inactive_wires)
+
+    # convert to computational basis
+    # i.e., converting the list of basis state bit strings into
+    # a list of decimal numbers that correspond to the computational
+    # basis state. For example, [0, 1, 0, 1, 1] = 2^3+2^1+2^0 = 11.
+    perm = np.ravel_multi_index(tuples[:, rearranged_wires].T, [2] * N)
+
+    # permute U to take into account rearranged wires
+    return U[:, perm][perm]
+
+
+def to_DiGraph(queue, observables):
+    """Convert a queue and observables to a directed acyclic graph.
+
+    The resulting graph has nodes representing quantum operations,
+    and edges representing dependent/successor operations.
+
+    Each node is labelled by an integer corresponding to the position
+    in the queue; note attributes are used to store information about the node:
+
+    * ``'name'`` *(str)*: name of the quantum operation (e.g., ``'PauliZ'``)
+
+    * ``'op'`` *(Operation or Observable)*: the quantum operation/observable object
+
+    Args:
+        queue (list[Operation]): the quantum operations to apply
+        observables (list[Observable]): the quantum observables to measure
+
+    Returns:
+        networkx.DiGraph: the directed acyclic graph representing
+        the quantum program
+    """
+    grid = {}
+
+    for idx, op in enumerate(queue + observables):
+        cmd = Command(name=op.name, op=op)
+
+        for q in set(op.wires):
+            # Add cmd to the grid to the end of the line r.ind.
+            if q not in grid:
+                # add a new line to the circuit
+                grid[q] = []
+
+            grid[q].append([idx, cmd])
+
+    G = nx.DiGraph()
+
+    for q, cmds in grid.items():
+        if cmds:
+            # add the first operation on the wire that does not depend on anything
+            attrs = cmds[0][1]._asdict()
+            G.add_node(cmds[0][0], **attrs)
+
+        for i in range(1, len(cmds)):
+            # add the edge between the operations, and the operation nodes themselves
+            if cmds[i][0] not in G:
```

Comment:
This ensures that subsequent operations on the wire are only added to the graph if not already on the graph. This will always be the case for single-wire operations.

Multi-wire operations might already be on the graph, so this just avoids adding them twice.